app/controllers/concerns/course/discussion/posts_concern.rb

```diff
@@ -0,0 +1,62 @@
+module Course::Discussion::PostsConcern
+  extend ActiveSupport::Concern
+
+  included do
+    before_action :set_topic
+    authorize_resource :topic
+    load_and_authorize_resource :post, through: :topic, class: Course::Discussion::Post.name
+  end
+
+  def create
+    Course::Discussion::Post.transaction do
+      fail ActiveRecord::Rollback unless @post.save
+      fail ActiveRecord::Rollback unless auto_subscribe_topic
+      true
+    end
+  end
+
+  def edit
+  end
+
+  def update
+    @post.update_attributes(post_params)
+  end
+
+  def destroy
+    @post.destroy
+  end
+
+  # Render a new post in a separate page
+  # This can be used when you do not want to introduce javascript to your implementation
+  def reply
+    @child_post = topic.posts.build(title: t('course.discussion.posts.reply_title',
+                                             title: @post.title))
+  end
+
+  protected
+
+  # Create topic subscriptions for related users
+  #
+  # @return [Bool] True if all subscriptions are created successfully.
+  def auto_subscribe_topic
+    fail NotImplementedError, 'To be implemented by the concrete topic posts controller.'
+  end
+
+  # The discussion topic record that posts belong to.
+  # When your model uses 'acts_as :topic', you can write: 'your_instance.topic' in this method.
+  #
+  # @return [Course::Discussion::Topic] The discussion topic record.
+  def topic
+    fail NotImplementedError, 'To be implemented by the concrete topic posts controller.'
+  end
+
+  private
+
+  def set_topic
+    @topic ||= topic
```

Comment:
This will rewrite `@topic` in Forum::PostController also. Because, in Forum::PostController, there is also an instance variable called `@topic`(concrete topic). How can I avoid this problem? Need some suggestions. @lowjoel @allenwq